pxr/usd/usdSemantics/labelsQuery.h

```diff
@@ -0,0 +1,129 @@
+//
+// Copyright 2024 Pixar
+//
+// Licensed under the Apache License, Version 2.0 (the "Apache License")
+// with the following modification; you may not use this file except in
+// compliance with the Apache License and the following modification to it:
+// Section 6. Trademarks. is deleted and replaced with:
+//
+// 6. Trademarks. This License does not grant permission to use the trade
+//    names, trademarks, service marks, or product names of the Licensor
+//    and its affiliates, except as required to comply with Section 4(c) of
+//    the License and to reproduce the content of the NOTICE file.
+//
+// You may obtain a copy of the Apache License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the Apache License with the above modification is
+// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. See the Apache License for the specific
+// language governing permissions and limitations under the Apache License.
+//
+
+#ifndef PXR_USDSEMANTICS_LABELSQUERY_H
+#define PXR_USDSEMANTICS_LABELSQUERY_H
+
+#include "pxr/pxr.h"
+
+#include "pxr/usd/usdSemantics/labelsAPI.h"
+#include "pxr/usd/usd/timeCode.h"
+#include "pxr/base/gf/interval.h"
+#include "pxr/base/tf/token.h"
+
+#include <shared_mutex>
+#include <set>
+#include <unordered_map>
+#include <variant>
+
+
+PXR_NAMESPACE_OPEN_SCOPE
+
+/// The `UsdSemanticsLabelsQuery` can be used to query a prim's
+/// labels for a specified taxonomy and time from the `UsdSemanticsLabelsAPI`.
+/// Time may be an individual time code or an interval.
+///
+/// This utility requires that all prims are on the same stage.
+///
+/// The query caches certain reads and computations and should be discarded
+/// when the state of the stage changes.
```

Comment:
Helpful to mention that it is safe to use in a multi-threaded context... not all caches in Usd schemas are currently, regrettably.